modules/apps/notification/notification-service/src/main/java/com/liferay/notification/internal/type/email/provider/TermEmailProvider.java

```diff
@@ -0,0 +1,130 @@
+import com.liferay.portal.kernel.security.permission.resource.ModelResourcePermissionUtil;
+import com.liferay.portal.kernel.service.RoleLocalService;
+import com.liferay.portal.kernel.service.UserLocalService;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author Carolina Barbosa
+ */
+public class TermEmailProvider implements EmailProvider {
+
+	public TermEmailProvider(
+		PermissionCheckerFactory permissionCheckerFactory,
+		NotificationTermEvaluatorTracker notificationTermEvaluatorTracker,
+		RoleEmailProvider roleEmailProvider, RoleLocalService roleLocalService,
+		UserLocalService userLocalService) {
+
+		_permissionCheckerFactory = permissionCheckerFactory;
+		_notificationTermEvaluatorTracker = notificationTermEvaluatorTracker;
+		_roleEmailProvider = roleEmailProvider;
+		_roleLocalService = roleLocalService;
+		_userLocalService = userLocalService;
+	}
+
+	@Override
+	public String provide(NotificationContext notificationContext, Object value)
+		throws PortalException {
+
+		if (value == null) {
+			return StringPool.BLANK;
+		}
+
+		Set<String> emailAddresses = new HashSet<>();
+
+		for (String term : StringUtil.split(GetterUtil.getString(value))) {
+			Matcher matcher = _termNamePattern.matcher(term);
+
+			if (!matcher.find()) {
+				continue;
+			}
+
+			for (NotificationTermEvaluator notificationTermEvaluator :
+					_notificationTermEvaluatorTracker.
+						getNotificationTermEvaluators(
+							notificationContext.getClassName())) {
+
+				String termValue = notificationTermEvaluator.evaluate(
+					NotificationTermEvaluator.Context.RECIPIENT,
+					notificationContext.getTermValues(), term);
+
+				if (Objects.equals(term, termValue)) {
+					continue;
+				}
+
+				matcher = _emailAddressPattern.matcher(termValue);
+
+				if (matcher.find()) {
+					emailAddresses.add(termValue);
+
+					continue;
+				}
+
+				Role role = _roleLocalService.fetchRole(
+					GetterUtil.getLong(termValue));
+
+				if (role != null) {
+					return _roleEmailProvider.provide(
+						notificationContext,
+						Collections.singletonList(
+							Collections.singletonMap(
+								NotificationRecipientSettingConstants.
+									NAME_ROLE_NAME,
+								role.getName())));
+				}
+
+				User user = _userLocalService.getUser(
+					GetterUtil.getLong(termValue));
+
+				if (ModelResourcePermissionUtil.contains(
+						_permissionCheckerFactory.create(user),
+						notificationContext.getGroupId(),
+						notificationContext.getClassName(),
+						notificationContext.getClassPK(), ActionKeys.VIEW)) {
+
+					emailAddresses.add(user.getEmailAddress());
+				}
+			}
+		}
+
+		return StringUtil.merge(emailAddresses);
+	}
+
+	private static final Pattern _emailAddressPattern = Pattern.compile(
```

Comment:
Can we reuse this existing [DefaultEmailProvider#_emailAddressPattern](https://github.com/liferay/liferay-portal/blob/c1098149eec3169865ccec75deb2c37e46bbfd37/modules/apps/notification/notification-service/src/main/java/com/liferay/notification/internal/type/email/provider/DefaultEmailProvider.java#L76) somehow?

Reply:
Done!